src/i2cint.rs

```diff
@@ -0,0 +1,386 @@
+                self.rx_ind += 1;
+            }
+            Ok(State::TxStop) => {
+                self.tx_ind = 0;
+                self.current_write_addr = None;
+            }
+            Ok(State::TxComplete) => {
+                self.tx_ind = 0;
+                // When receiving Tx complete, we should read after
+                // if not it should be a tx stop
+                if let Some(recv) = self.recv {
+                    self.read(recv.0, recv.1)
+                } else {
+                    self.last_error = Some(I2cError::TransferCompleteNoRead);
+                }
+                self.current_write_addr = None;
+            }
+            Ok(State::RxStop) => {
+                self.rx_ind = 0;
+            }
+            Ok(State::TxNack) => {
+                self.tx_ind = 0;
+                self.last_error = Some(I2cError::Nack);
+                self.current_write_addr = None;
+            }
+            Err(err) => {
+                self.last_error = Some(err);
+                self.current_write_addr = None;
+            }
+            _ => {}
+        }
+    }
+
+    /// Computes the states based on IRS register.
+    fn isr_state(&mut self) -> Result<State, I2cError> {
+        let isr = self.dev.i2c.isr.read();
+        if isr.berr().bit() {
+            self.dev.i2c.icr.write(|w| w.berrcf().bit(true));
+            return Err(I2cError::BusError);
+        } else if isr.arlo().bit() {
+            self.dev.i2c.icr.write(|w| w.arlocf().bit(true));
+            return Err(I2cError::ArbitrationLoss);
+        } else if isr.ovr().bit() {
+            self.dev.i2c.icr.write(|w| w.ovrcf().bit(true));
+            return Err(I2cError::Overrun);
+        }
+        if isr.nackf().bit() {
+            self.dev.i2c.icr.write(|w| w.nackcf().bit(true));
+            self.state = State::TxNack;
+        } else if isr.tc().bit() {
+            self.state = State::TxComplete;
+        } else if isr.txis().bit() && isr.txe().bit() {
+            self.state = State::TxReady;
+        } else if isr.rxne().bit() {
+            self.state = State::RxReady;
+        } else if isr.stopf().bit() {
+            // clear stop bit once read
+            self.dev.i2c.icr.write(|w| w.stopcf().bit(true));
+            self.state = match self.state {
+                State::TxSent | State::TxComplete => State::TxStop,
+                State::RxReady => State::RxStop,
+                _ => return Err(I2cError::StateError),
+            }
+        }
+        Ok(self.state)
+    }
+
+    /// Write a bute into the tx buffer.
+    fn write_tx_buffer(&mut self, byte: u8) {
+        self.dev.i2c.txdr.write(|w| w.txdata().bits(byte));
+        self.state = State::TxSent;
+    }
+
+    /// Is the device Busy ?
+    /// # Returns
+    /// true if busy, false if not.
+    pub fn is_busy(&mut self) -> bool {
+        self.dev.i2c.isr.read().busy().bit()
+    }
+
+    /// Get the state of the device.
+    pub fn get_tx_state(&self) -> State {
```

Comment:
`State` does include RX and TX states. Therefor this function name does not make sense?

```suggestion
    pub fn get_state(&self) -> State {
```